### DISCUSSION TITLE
Focus behavior when opening a dialog

### DISCUSSION THREAD

**Discussion by irangarcia** [ariakit/ariakit#1667] [Questions] [answered]
On this [example](https://ariakit.org/examples/menu-dialog-animated) after opening the menu and clicking on the `Create list` option, the Dismiss button doesn't get focused. 

But if I select the `Create list` option hitting `enter`, the dismiss button get focused, is this the expected behavior?

**Comment by diegohaz** [comment, marked as answer]
I'm sorry. I misunderstood your question. See https://github.com/ariakit/ariakit/issues/1384#issuecomment-1132540411

---

Old answer:

The menu has special behavior. That's expected:
https://github.com/ariakit/ariakit/blob/b1f3d9b3bcab322494e4460c7e377e222f5cbf2b/packages/ariakit/src/menu/menu-button.ts#L131

You can override it, though:

```jsx
const menu = useMenuState();

// Directly in the render phase. No need to put in an effect.
if (menu.initialFocus === "container") {
  menu.setInitialFocus("first")
}
```

**Reply by irangarcia** [reply]
Oh, thanks. 

I thought this could have been a wrong behavior on the Dialog from this [PR](https://github.com/ariakit/ariakit/pull/1422), I asked that because I'm having a problem about focusing when opening the dialog on a jest environment, as you can see on this [CodeSandbox](https://codesandbox.io/s/interesting-water-yorml4?file=/src/index.test.tsx). Before the release, the test was working just fine.

**Reply by diegohaz** [reply]
JSDOM doesn't support CSS animations. That's why we use [Playwright](https://github.com/ariakit/ariakit/blob/main/contributing.md#writing-end-to-end-tests) to test those examples.

The dialog only gets focused after the enter animation completes [by listening to the `transitionend`/`animationend` events](https://github.com/ariakit/ariakit/blob/6646b70f3a6cd8809645a792d7f2618455b2333b/packages/ariakit/src/disclosure/disclosure-content.ts#L60-L84). These events are never fired in JSDOM.

I'm not sure why your test was working before.

**Reply by irangarcia** [reply]
Thank you for the answer. 

We were able to create a workaround for this.

```tsx
it("should initially focus on the specified target reference", async () => {
    const ref = createRef<HTMLInputElement>();

    render(
      <Dialog initialFocus={ref}>
        <input ref={ref} type="input" />
        <p>Lorem ipsum</p>
      </Dialog>
    );

    await expect(screen.findByRole('dialog', {hidden: false})).resolves.toBeInTheDocument();

    fireEvent.animationEnd(screen.getByRole('dialog'));

    expect(ref.current).toHaveFocus();
});
```
